src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp

```diff
@@ -0,0 +1,200 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "gc/shenandoah/shenandoahHeap.inline.hpp"
+#include "gc/shenandoah/shenandoahHeapRegion.hpp"
+#include "gc/shenandoah/shenandoahUncommitThread.hpp"
+#include "logging/log.hpp"
+#include "runtime/mutexLocker.hpp"
+#include "utilities/events.hpp"
+
+ShenandoahUncommitThread::ShenandoahUncommitThread(ShenandoahHeap* heap)
+  : _heap(heap),
+    _stop_lock(Mutex::safepoint - 2, "ShenandoahUncommitStop_lock", true),
+    _uncommit_lock(Mutex::safepoint - 2, "ShenandoahUncommitCancel_lock", true) {
+  set_name("Shenandoah Uncommit Thread");
+  create_and_start();
+
+  // Allow uncommits. This is managed by the control thread during a GC.
+  _uncommit_allowed.set();
+}
+
+void ShenandoahUncommitThread::run_service() {
+  assert(ShenandoahUncommit, "Thread should only run when uncommit is enabled");
+
+  // Shrink period avoids constantly polling regions for shrinking.
+  // Having a period 10x lower than the delay would mean we hit the
+  // shrinking with lag of less than 1/10-th of true delay.
+  // ShenandoahUncommitDelay is in millis, but shrink_period is in seconds.
+  double shrink_period = (double)ShenandoahUncommitDelay / 1000 / 10;
+  double last_shrink_time = os::elapsedTime();
+  while (!should_terminate()) {
+    double current = os::elapsedTime();
+    bool soft_max_changed = _soft_max_changed.try_unset();
+    bool explicit_gc_requested = _explicit_gc_requested.try_unset();
+
+    if (soft_max_changed || explicit_gc_requested || current - last_shrink_time > shrink_period) {
+      size_t shrink_until = soft_max_changed ? _heap->soft_max_capacity() : _heap->min_capacity();
+      double shrink_before = (soft_max_changed || explicit_gc_requested) ?
+              current :
+              current - ((double) ShenandoahUncommitDelay / 1000.0);
+
+      // Explicit GC tries to uncommit everything down to min capacity.
+      // Soft max change tries to uncommit everything down to target capacity.
+      // Periodic uncommit tries to uncommit suitable regions down to min capacity.
+      if (should_uncommit(shrink_before, shrink_until)) {
+        uncommit(shrink_before, shrink_until);
+        last_shrink_time = current;
+      }
+    }
+    {
+      MonitorLocker locker(&_stop_lock, Mutex::_no_safepoint_check_flag);
+      if (!_stop_requested.is_set()) {
+        locker.wait((int64_t)shrink_period);
```

Comment:
I tried to test this on some of my toy examples, and realized this particular line may end up as `locker.wait(0)`, which means "wait indefinitely, until notified". This breaks periodic commits. The old code rode on control thread doing `MAX2(1, ...)`, so we never feed `0` into `wait`. I am also confused about units. The comment above says `shrink_period` is in seconds, but `locker.wait` accepts milliseconds?

Reply:
It sounds like this line should be:

```
        locker.wait(MAX2<int64_t>(1, shrink_period * 1000));
```

Reply:
Sorry, I missed your comments here. I noticed the same and have refactored this code.